M2/Macaulay2/packages/Colon.m2

```diff
@@ -0,0 +1,885 @@
+--        5. employ storefun! Maybe call it declare
+---------------------------------------------------------------------------
+newPackage(
+    "Colon",
+    Version => "0.2",
+    Date => "November 8, 2020",
+    Headline => "quotient, saturation, and annihilator routines for ideals and modules",
+    Authors => {
+	{Name => "Justin Chen",    Email => "justin.chen@math.gatech.edu"},
+	{Name => "Mahrud Sayrafi", Email => "mahrud@umn.edu",        HomePage => "https://math.umn.edu/~mahrud"},
+	{Name => "Mike Stillman",  Email => "mike@math.cornell.edu", HomePage => "http://www.math.cornell.edu/~mike"}},
+    Keywords => {"Commutative Algebra"},
+    PackageExports => { "Elimination" },
+    AuxiliaryFiles => true,
+    DebuggingMode => false
+    )
+
+export { "isSupportedInZeroLocus" }
+
+exportFrom_Core { "saturate", "annihilator" }
+
+importFrom_Core { "printerr", "raw", "rawColon", "rawSaturate", "newMonomialIdeal", "eliminationInfo" }
+
+-- TODO: where should these be placed?
+trim MonomialIdeal := MonomialIdeal => opts -> (cacheValue (symbol trim => opts)) ((I) -> monomialIdeal trim(module I, opts))
+
+-- TODO: is this the right function?
+ambient Ideal := Ideal => I -> ideal 1_(ring I)
+
+--Ideal % Matrix            :=
+--remainder(Ideal,  Matrix) := Matrix => (I, m) -> remainder(gens I, m)
+--Module % Matrix           :=
+--remainder(Module, Matrix) := Matrix => (M, m) -> remainder(gens M, m)
+
+-- This is a map from method keys to strategy hash tables
+algorithms := new MutableHashTable from {}
+
+--------------------------------------------------------------------
+-- Helpers
+--------------------------------------------------------------------
+
+cacheHit := type -> if debugLevel > 0 then printerr("Cache hit on a ", synonym type, "! 🎉");
+
+removeOptions := (opts, badopts) -> (
+    opts = new MutableHashTable from opts;
+    scan(badopts, k -> remove(opts, k));
+    new OptionTable from opts)
+
+removeQuotientOptions := opts -> (
+    opts = new MutableHashTable from opts;
+    remove(opts, Strategy);
+    remove(opts, MinimalGenerators);
+    -- TODO: what would this do?
+    --opts.SyzygyLimit = opts.BasisElementLimit;
+    --remove(opts,BasisElementLimit);
+    new OptionTable from opts)
+
+-- given a ring R, determines if R is a poly ring over ZZ or a field
+isFlatPolynomialRing = R -> isPolynomialRing R and (isField(kk := coefficientRing R) or kk === ZZ)
+
+-- TODO: can this work with multigraded ideals?
+isGRevLexRing = (R) -> (
+     -- returns true if the monomial order in the polynomial ring R
+     -- is graded reverse lexicographic order, w.r.t. the first degree
+     -- vector in the ring.
+     mo := (options monoid R).MonomialOrder;
+     mo = select(mo, x -> x#0 =!= MonomialSize and x#0 =!= Position);
+     isgrevlex := mo#0#0 === GRevLex and mo#0#1 === apply(degrees R, first);
+#mo === 1 and isgrevlex and all(mo, x -> x#0 =!= Weights and x#0 =!= Lex))
+
+-- Helper for Linear strategies
+isLinearForm := f -> (
```

Comment:
The name is fine, especially since it isn't exported, but I think it doesn't support multigradings currently. I'll add a TODO item, but don't wait on it, because we want this package to go out in this release but this fix might have to wait for the next release.